### ISSUE TITLE
any hints to compile with GCC on windows 10?

### ISSUE THREAD

**Issue by retsyo** [shedskin/shedskin#342]
I can use old shedskin to compile python 2 code in MSYS2/Mingw64 on windows 10.

however, the latest shedskin can't compile even
```python
'''test.py'''
print('hello, world!')
```
1. command
```
/e/prg/py/winpython/python-3.8.1.amd64/python /e/prg/py/winpython/python-3.8.1.amd64/Scripts/shedskin test
```
can generate `test.hpp` and `Makefile`

2. however `execinfo.h` is not found
```
 make
g++  -O2 -std=c++17 -march=native -Wno-deprecated -Wl,--enable-auto-import  -O3 -static-libgcc -Wl,--kill-at -static -I. -IE:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib C:\tmp\shedskin-master\examples\test.cpp E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.cpp -lgc -lpcre  -o test
In file included from C:\tmp\shedskin-master\examples\test.cpp:1:
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib/builtin.hpp:34:10: fatal error: execinfo.h: No such file or directory
   34 | #include <execinfo.h>
      |          ^~~~~~~~~~~~
compilation terminated.
In file included from E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.cpp:3:
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.hpp:34:10: fatal error: execinfo.h: No such file or directory
   34 | #include <execinfo.h>
      |          ^~~~~~~~~~~~
compilation terminated.
make: *** [Makefile:15: test] Error 1

```

so I modified `e:\prg\py\winpython\python-3.8.1.amd64\Lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.hpp`
from 
```cpp
#ifndef WIN32
#include <cxxabi.h>
#include <exception>
#ifndef __APPLE__
#include <execinfo.h>
#endif
#endif
```

to
```cpp
#ifndef WIN32
#include <cxxabi.h>
#include <exception>
//#ifndef __APPLE__
//#include <execinfo.h>
//#endif
#endif
``` 

3. `make` again, but get
```bash
g++  -O2 -std=c++17 -march=native -Wno-deprecated -Wl,--enable-auto-import  -O3 -static-libgcc -Wl,--kill-at -static -I. -IE:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib C:\tmp\shedskin-master\examples\test.cpp E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.cpp -lgc -lpcre  -o test
In file included from E:/msys64/mingw64/include/c++/12.1.0/cctype:42,
                 from E:/msys64/mingw64/include/c++/12.1.0/bits/localefwd.h:42,
                 from E:/msys64/mingw64/include/c++/12.1.0/string:43,
                 from E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib/builtin.hpp:15,
                 from C:\tmp\shedskin-master\examples\test.cpp:1:
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib/builtin.hpp:397:15: error: expected unqualified-id before 'unsigned'
  397 |     __ss_bool isascii();
      |               ^~~~~~~
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib/builtin.hpp:397:15: error: expected ')' before 'unsigned'
  397 |     __ss_bool isascii();
      |               ^~~~~~~
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib/builtin.hpp:397:15: error: expected ')' before 'unsigned'
  397 |     __ss_bool isascii();
      |               ^~~~~~~
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib/builtin.hpp:560:15: error: expected unqualified-id before 'unsigned'
  560 |     __ss_bool isascii();
      |               ^~~~~~~
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib/builtin.hpp:560:15: error: expected ')' before 'unsigned'
  560 |     __ss_bool isascii();
      |               ^~~~~~~
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib/builtin.hpp:560:15: error: expected ')' before 'unsigned'
  560 |     __ss_bool isascii();
      |               ^~~~~~~
In file included from E:/msys64/mingw64/include/c++/12.1.0/cctype:42,
                 from E:/msys64/mingw64/include/c++/12.1.0/bits/localefwd.h:42,
                 from E:/msys64/mingw64/include/c++/12.1.0/string:43,
                 from E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.hpp:15,
                 from E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.cpp:3:
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.hpp:397:15: error: expected unqualified-id before 'unsigned'
  397 |     __ss_bool isascii();
      |               ^~~~~~~
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.hpp:397:15: error: expected ')' before 'unsigned'
  397 |     __ss_bool isascii();
      |               ^~~~~~~
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.hpp:397:15: error: expected ')' before 'unsigned'
  397 |     __ss_bool isascii();
      |               ^~~~~~~
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.hpp:560:15: error: expected unqualified-id before 'unsigned'
  560 |     __ss_bool isascii();
      |               ^~~~~~~
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.hpp:560:15: error: expected ')' before 'unsigned'
  560 |     __ss_bool isascii();
      |               ^~~~~~~
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.hpp:560:15: error: expected ')' before 'unsigned'
  560 |     __ss_bool isascii();
      |               ^~~~~~~
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/str.cpp:141:16: error: expected unqualified-id before '(' token
  141 | __ss_bool str::isascii() {
      |                ^~~~~~~
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/bytes.cpp:549:18: error: expected unqualified-id before '(' token
  549 | __ss_bool bytes::isascii() {
      |                  ^~~~~~~
In file included from E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.cpp:169:
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp: In function '__shedskin__::str* __shedskin__::do_asprintf_str(const char*, str*, pyobj*, pyobj*)':
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp:95:13: error: 'asprintf' was not declared in this scope; did you mean 'vsprintf'?
   95 |         x = asprintf(&d, fmt, ((int)(((int_ *)a1)->unit)), ((int)(((int_ *)a2)->unit)), s->c_str());
      |             ^~~~~~~~
      |             vsprintf
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp:97:13: error: 'asprintf' was not declared in this scope; did you mean 'vsprintf'?
   97 |         x = asprintf(&d, fmt, ((int)(((int_ *)a1)->unit)), s->c_str());
      |             ^~~~~~~~
      |             vsprintf
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp:99:13: error: 'asprintf' was not declared in this scope; did you mean 'vsprintf'?
   99 |         x = asprintf(&d, fmt, s->c_str());
      |             ^~~~~~~~
      |             vsprintf
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp: In instantiation of '__shedskin__::str* __shedskin__::do_asprintf(const char*, T, pyobj*, pyobj*) [with T = int]':
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp:173:26:   required from here
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp:70:21: error: 'asprintf' was not declared in this scope; did you mean 'vsprintf'?
   70 |         x = asprintf(&d, fmt, ((int)(((int_ *)a1)->unit)), ((int)(((int_ *)a2)->unit)), t);
      |             ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |             vsprintf
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp:72:21: error: 'asprintf' was not declared in this scope; did you mean 'vsprintf'?
   72 |         x = asprintf(&d, fmt, ((int)(((int_ *)a1)->unit)), t);
      |             ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |             vsprintf
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp:74:21: error: 'asprintf' was not declared in this scope; did you mean 'vsprintf'?
   74 |         x = asprintf(&d, fmt, t);
      |             ~~~~~~~~^~~~~~~~~~~~
      |             vsprintf
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp: In instantiation of '__shedskin__::str* __shedskin__::do_asprintf(const char*, T, pyobj*, pyobj*) [with T = double]':
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp:180:26:   required from here
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp:70:21: error: 'asprintf' was not declared in this scope; did you mean 'vsprintf'?
   70 |         x = asprintf(&d, fmt, ((int)(((int_ *)a1)->unit)), ((int)(((int_ *)a2)->unit)), t);
      |             ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |             vsprintf
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp:72:21: error: 'asprintf' was not declared in this scope; did you mean 'vsprintf'?
   72 |         x = asprintf(&d, fmt, ((int)(((int_ *)a1)->unit)), t);
      |             ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |             vsprintf
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin/format.cpp:74:21: error: 'asprintf' was not declared in this scope; did you mean 'vsprintf'?
   74 |         x = asprintf(&d, fmt, t);
      |             ~~~~~~~~^~~~~~~~~~~~
      |             vsprintf
make: *** [Makefile:15: test] Error 1

```

so any fix? Thanks

**Comment by srepmub** [comment]
sorry about this. I have not tested shedskin 0.9.5/0.9.6 on windows before releasing. I will have a look over the weekend..

the missing asprintf error is weird, as format.cpp defines it itself on windows.. the other errors may be because shedskin now uses c++17, not sure. that may give other problems with an old shedskin-mingw-tarball in any case..

note that @shakfu is working on cmake-support for shedskin, which sounds like a better approach than using mingw.

btw if you just want to compile something locally, you could try to use WSL.

**Comment by shakfu** [comment]
Yes as @srepmub has said, the priority was getting the unix-like OSes (linux, osx) working for the release, but there was definitely some preparation for working on windows support  in the future:

- [cmake](https://cmake.org): there was comprehensive support of building all shedskin tests using cmake (see the `tests` directory for how this is done: root and subdirectories `CMakeLists.txt` and cmake scripts and functions in the `cmake` directory).

- [conan](https://conan.io): investigation and windows test of the conan cross-platform c/c++ pkg manager which works well with cmake. The initial test on windows was able to build and install shedskin dependencies on windows (see [this] (https://github.com/shedskin/shedskin/issues/324#issuecomment-1366347498) post for details)

So basically the strategy is clear: use an appropriate windows configuration with cmake and conan, and it is definitely on the todo list.

Of course, it would be great if someone with a windows machine can participate in this project and help with developing and testing windows support (-:

**Comment by srepmub** [comment]
I committed some minor changes. let me know if this fixes the 'hello, world' test at least =)

so one issue is that apparently "WIN32" is not defined anymore in Mingw64.. but "_WIN32" is. sigh. another is that 'isascii' is (now?) some kind of macro included somewhere, so I added it to shedskin/illegal (so it becomes __ss_isascii).

we check for WIN32 in 100 places, so perhaps we can just define it ourselves in the Makefile? see here:

https://stackoverflow.com/questions/9025708/mingw-not-defining-win32-error-in-preprocessor-directives

**Comment by carlkl** [comment]
The latest [commit](https://github.com/shedskin/shedskin/commit/9e4c04e79f9f539708c6fb73f6655a898536150c) is fine. Only two things are still missing for Windows: 

1. A patch for [builtin.hpp](https://github.com/shedskin/shedskin/blob/master/shedskin/lib/builtin.hpp) to not include `execinfo.h` as described above 
2. adding `-lgccpp` to [FLAGS.mingw](https://github.com/shedskin/shedskin/blob/master/shedskin/FLAGS.mingw)

With that and with a proper configured MSYS2 environment (mingw32, mingw42, ucrt64) it is possible to compile and link most of the example programs.

**Comment by retsyo** [comment]
thanks, now `hello.exe` can be created


1. 

> The latest [commit](https://github.com/shedskin/shedskin/commit/9e4c04e79f9f539708c6fb73f6655a898536150c) is fine. Only two things are still missing for Windows:
> 
> 1. A patch for [builtin.hpp](https://github.com/shedskin/shedskin/blob/master/shedskin/lib/builtin.hpp) to not include `execinfo.h` as described above
> 2. adding `-lgccpp` to [FLAGS.mingw](https://github.com/shedskin/shedskin/blob/master/shedskin/FLAGS.mingw)
> 
> With that and with a proper configured MSYS2 environment (mingw32, mingw42, ucrt64) it is possible to compile and link most of the example programs.

**Comment by retsyo** [comment]
> btw if you just want to compile something locally, you could try to use WSL.
so bad, there is not so much free disk storage on my notebook to install WSL :(

**Comment by srepmub** [comment]
thanks guys! I will update the flags and improve the WIN32 situation (so execinfo.h is also not included) hopefully later today.

@retsyo: no WSL is really just a convenient way to use linux. I also don't know what the cross-compilations options are.

**Comment by srepmub** [comment]
let's keep this open a bit longer. we probably also want to give a short description in the docs of how to use shedskin master on windows, before the 0.9.7 release.

**Comment by retsyo** [comment]
well, I pulled request on `FLAGS.mingw` and `lib\time.cpp` to fix compilation problem with mingw

TLDR:
I found that if we add `-DWIN32` for `CCFLAGS`, then we can just leave `builtin.hpp` unchanged

further more,  `-DWIN32`  erase 
```
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\time.cpp: In function '__time__::struct_time* __time__::strptime(__shedskin__::str*, __shedskin__::str*)':
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\time.cpp:892:54: error: too many initializers for 'tm'
  892 |     tm time_tuple = {0, 0, 0, 1, 0, 0, 0, 1, -1, 0, 0};
      |                                                      ^
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\time.cpp:893:11: error: '::strptime' has not been declared; did you mean '__time__::strptime'?
  893 |     if(!::strptime(string->c_str(), format->c_str(), &time_tuple))
      |           ^~~~~~~~
      |           __time__::strptime
E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\time.cpp:885:14: note: '__time__::strptime' declared here
  885 | struct_time *strptime(str *string, str *format) {
      |              ^~~~~~~~
make: *** [Makefile:23: amaze] Interrupt
```

**Comment by srepmub** [comment]
makes sense, yeah. all merged, thank you! =)

did you just copy the pcre, libgc dlls and libgc includes from shedskin 0.9.4..?

**Comment by retsyo** [comment]
no, I have installed pcre and libgc in Msys2/mingw64 a long time ago via `pacman -S something`

**Comment by srepmub** [comment]
oh, that's nice. would you or anyone be willing to provide full instructions to get shedskin 0.9.6 working like this? (it sounds like it's just a matter of install msys2, mingw64, shedskin (0.9.6 just turned up on pypi btw), and via pacman libgc and libpcre..?). ideally in the form of a PR on the docs/ directory. I can also go through the process myself once.

**Comment by srepmub** [comment]
ehm well of course 0.9.6 won't work directly, but master then..

**Comment by retsyo** [comment]
yes, just install msys2, mingw64, lib(libpcre via pacman, libgc via `configure` and `make install`.  Sorry I don't remember every lib exactly because I installed such libs a long time ago), shedskin as ususal. Then almost all example codes can be compiled. So I think there is no need to write a specified document

however, there are some exceptions. for example, `examples\minilight.py` for which I have not find a solution yet:

```bash
$ python3 -m shedskin minilight
*** SHED SKIN Python-to-C++ Compiler 0.9.7 *** -
Copyright 2005-2022 Mark Dufour and contributors; License GNU GPL version 3 (See LICENSE)

|                           |  0%
|**********                 | 37%
|*************************  | 94%
|***************************|100%

*WARNING* ml\image.py:52: 'isinstance' is not supported; always returns True
*WARNING* ml\minilight.py:96: system 'KeyboardInterrupt' is not caught
*WARNING* ml\spatialindex.py:74: 'isinstance' is not supported; always returns True
*WARNING* ml\surfacepoint.py:21: 'isinstance' is not supported; always returns True
[elapsed time: 4.84 seconds]

```  
then
```
$ make
g++  -O2 -DWIN32 -std=c++17 -march=native -Wno-deprecated -ffunction-sections -fdata-sections -Wl,-gc-sections -Wl,-s -Wl,--enable-auto-import  -O3 -static-libgcc -Wl,--kill-at -static -I. -IE:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib C:\tmp\shedskin-master\examples\ml\vector3f.cpp C:\tmp\shedskin-master\examples\ml\triangle.cpp C:\tmp\shedskin-master\examples\ml\surfacepoint.cpp C:\tmp\shedskin-master\examples\ml\spatialindex.cpp C:\tmp\shedskin-master\examples\ml\scene.cpp C:\tmp\shedskin-master\examples\ml\raytracer.cpp C:\tmp\shedskin-master\examples\ml\minilight.cpp C:\tmp\shedskin-master\examples\ml\image.cpp C:\tmp\shedskin-master\examples\ml\camera.cpp C:\tmp\shedskin-master\examples\ml\__init__.cpp C:\tmp\shedskin-master\examples\minilight.cpp E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\time.cpp E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\sys.cpp E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\re.cpp E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\random.cpp E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\math.cpp E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\builtin.cpp -lgc -lpcre -lgccpp  -lpcre -o minilight
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:\msys64\tmp\cc83gJD4.o:re.cpp:(.text$_ZN6__re__6__initEv+0x2a): undefined reference to `__imp_pcre_malloc'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:\msys64\tmp\cc83gJD4.o:re.cpp:(.text$_ZN6__re__6__initEv+0x3b): undefined reference to `__imp_pcre_free'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:\msys64\tmp\cc83gJD4.o:re.cpp:(.text$_ZN6__re__6__initEv+0x52): undefined reference to `__imp_pcre_maketables'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:\msys64\tmp\cc83gJD4.o:re.cpp:(.text$_ZN6__re__9re_object6__execEPN12__shedskin__3strEiii+0xa6): undefined reference to `__imp_pcre_exec'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:\msys64\tmp\cc83gJD4.o:re.cpp:(.text$_ZN6__re__9re_object6__subnEPN12__shedskin__3strES3_iPi+0xf1): undefined reference to `__imp_pcre_exec'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:\msys64\tmp\cc83gJD4.o:re.cpp:(.text$_ZN6__re__7compileEPN12__shedskin__3strEi+0xa7): undefined reference to `__imp_pcre_compile'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:\msys64\tmp\cc83gJD4.o:re.cpp:(.text$_ZN6__re__7compileEPN12__shedskin__3strEi+0x114): undefined reference to `__imp_pcre_study'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:\msys64\tmp\cc83gJD4.o:re.cpp:(.text$_ZN6__re__7compileEPN12__shedskin__3strEi+0x1b8): undefined reference to `__imp_pcre_fullinfo'
E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:\msys64\tmp\cc83gJD4.o:re.cpp:(.text$_ZN6__re__9re_object11__splitfindEPN12__shedskin__3strEici+0xf7): undefined reference to `__imp_pcre_exec'
collect2.exe: error: ld returned 1 exit status
make: *** [Makefile:45: minilight] Error 1

```

**Comment by srepmub** [comment]
okay, I will go over the steps again, and have a quick look at minilight.

**Comment by retsyo** [comment]
I tested the bundled examples in msys2/mingw64 on windows 10 by using `build_check_all.sh`. Most applications are speeded up as expected, and most `pyd` extensions are created as expected, which can be read in the last table 3 and 4. 

But let me address the failed examples in table 1 and 2:

table 1: failed pure application

|                 | pure python code | shedskin                                                     |
| --------------- | ---------------- | ------------------------------------------------------------ |
| minilight       | 35.305s          | undefined reference to `__imp_pcre_malloc'                   |
| othello         | 4.552s           | 20.218s. Yes, the `exe` runs slower then python dramaticaly. |
| rdb             | 0.252s           | undefined reference to `__imp_pcre_malloc'                   |
| sunfish         | 0.238s           | 'gbk' codec can't decode byte 0x9c in position 1602: illegal multibyte sequence |
| TonyJpegDecoder | 0.889s           | 1.048s. Yes, the `exe` runs slower then python a little.     |
| WebServer       |                  | E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\select.hpp:26:5: error: 'fd_set' was not declared in this scope |



table 2: failed extension 

| name       | message                                                      |
| ---------- | ------------------------------------------------------------ |
| c64        | E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:\msys64\tmp\cc4NzHs9.o:prg.cpp:(.text+0x6f): undefined reference to `__os__::__path__::basename(__shedskin__::bytes*)' |
| mastermind | undefined reference to `__imp_pcre_malloc'                   |





TL;DR

table 3: pure applications

  

| name            | pure python code | shedskin                                                     | py/shedskin |
| --------------- | ---------------- | ------------------------------------------------------------ | ----------- |
| ac_encode       | 0.414s           | 0.257s                                                       | 1.61        |
| adatron         | 36.757s          | 0.809s                                                       | 45.44       |
| amaze           | 3.651s           | 2.459s                                                       | 1.48        |
| ant             | 11.407s          | 0.588s                                                       | 19.40       |
| bh              | 18.532s          | 0.557s                                                       | 33.27       |
| block           | 0.541s           | 0.253s                                                       | 2.14        |
| brainfuck       | 1.046s           | 0.127s                                                       | 8.24        |
| chaos           | 8.927s           | 2.137s                                                       | 4.18        |
| chess           | 1.600s           | 0.191s                                                       | 8.38        |
| chull           | 0.959s           | 0.147s                                                       | 6.52        |
| dijkstra        | 0.924s           | 0.160s                                                       | 5.78        |
| dijkstra2       | 2.220s           | 0.828s                                                       | 2.68        |
| genetic         | 0.806s           | 0.126s                                                       | 6.40        |
| genetic2        | 37.497s          | 6.596s                                                       | 5.68        |
| go              | 26.634s          | 0.838s                                                       | 31.78       |
| hq2x            | 3.221s           | 0.172s                                                       | 18.73       |
| kanoodle        | 2.635s           | 0.414s                                                       | 6.36        |
| kmeanspp        | 6.146s           | 0.252s                                                       | 24.39       |
| life            | 0.490s           | 0.366s                                                       | 1.34        |
| linalg          | 1.319s           | 0.221s                                                       | 5.97        |
| loop            | 2.270s           | 0.550s                                                       | 4.13        |
| LZ2             | 0.510s           | 0.287s                                                       | 1.78        |
| mandelbrot      | 5.797s           | 0.258s                                                       | 22.47       |
| mao             | 14.062s          | 0.319s                                                       | 44.08       |
| mastermind2     | 0.747s           | 0.127s                                                       | 5.88        |
| minilight       | 35.305s          | undefined reference to `__imp_pcre_malloc'                   | NAN         |
| minpng          | 1.013s           | 0.338s                                                       | 3.00        |
| msp_ss          | 0.445s           | 0.023s                                                       | 19.35       |
| mwmatching      | 5.600s           | 0.460s                                                       | 12.17       |
| nbody           | 13.830s          | 0.326s                                                       | 42.42       |
| neural1         | 0.255s           | 0.109s                                                       | 2.34        |
| neural2         | 0.650s           | 0.128s                                                       | 5.08        |
| oliva2          | 2.056s           | 0.197s                                                       | 10.44       |
| othello         | 4.552s           | 20.218s                                                      | 0.23        |
| path_tracing    | 1m34.873s        | 2.704s                                                       | 35.09       |
| pisang          | 55.993s          | 1.831s                                                       | 30.58       |
| plcfrs          | 0.239s           | 0.112s                                                       | 2.13        |
| pygmy           | 1m2.594s         | 3.924s                                                       | 15.95       |
| pystone         | 0.455s           | 0.112s                                                       | 4.06        |
| quameon         | 1.587s           | 0.366s                                                       | 4.34        |
| rdb             | 0.252s           | undefined reference to `__imp_pcre_malloc'                   | NAN         |
| richards        | 1.132s           | 0.127s                                                       | 8.91        |
| rsync           | 0.257s           | 0.142s                                                       | 1.81        |
| rubik           | 3.815s           | 0.988s                                                       | 3.86        |
| rubik2          | 5m54.275s        | 30.342s                                                      | 11.68       |
| sat             | 2.352s           | 0.220s                                                       | 10.69       |
| score4          | 59.211s          | 1.024s                                                       | 57.82       |
| sha             | 0.236s           | 0.109s                                                       | 2.17        |
| sieve           | 1.911s           | 0.216s                                                       | 8.85        |
| sokoban         | 1.683s           | 0.397s                                                       | 4.24        |
| solitaire       | 0.254s           | 0.097s                                                       | 2.62        |
| sudoku1         | 3.891s           | 0.189s                                                       | 20.59       |
| sudoku2         | 0.256s           | 0.113s                                                       | 2.27        |
| sudoku3         | 1.225s           | 0.677s                                                       | 1.81        |
| sudoku4         | 1.907s           | 0.651s                                                       | 2.93        |
| sudoku5         | 0.241s           | 0.112s                                                       | 2.15        |
| sunfish         | 0.238s           | gbk' codec can't decode byte 0x9c in position 1602: illegal multibyte sequence | NAN         |
| TarsaLZP        | 0.239s           | 0.112s                                                       | 2.13        |
| tictactoe       | 0.238s           | 0.111s                                                       | 2.14        |
| timsort         | 10.854s          | 0.346s                                                       | 31.37       |
| TonyJpegDecoder | 0.889s           | 1.048s                                                       | 0.85        |
| voronoi         | 0.270s           | 0.110s                                                       | 2.45        |
| voronoi2        | 0.240s           | 0.126s                                                       | 1.90        |
| yopyra          | 9.145s           | 0.363s                                                       | 25.19       |
| WebServer       |                  | E:\prg\py\winpython\python-3.8.1.amd64\lib\site-packages\shedskin-0.9.7-py3.8.egg\shedskin\lib\select.hpp:26:5: error: 'fd_set' was not declared in this scope | NAN         |

table 4: `pyd` extensions

| name        | if failed, what is the message                               |
| ----------- | ------------------------------------------------------------ |
| astar       |                                                              |
| c64         | `E:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: E:\msys64\tmp\cc4NzHs9.o:prg.cpp:(.text+0x6f): undefined reference to `__os__::__path__::basename(__shedskin__::bytes*)'` |
| circle      |                                                              |
| fysphun     |                                                              |
| pylife      |                                                              |
| mandelbrot2 |                                                              |
| mastermind  | undefined reference to `__imp_pcre_malloc'                   |
| pygasus     |                                                              |
| pylot       |                                                              |
| stereo      |                                                              |

**Comment by srepmub** [comment]
not bad! =)

I managed to get shedskin running from scratch via msys64, using pacman and pip (pacman -S python-pip).. but for the minilight the pcre error does not occur for me (I do get other, time-related errors :S). I may have a look at those again.

all this is not really straightforward though.. :P

**Comment by srepmub** [comment]
are you sure you have libpcre installed, and not just pcre-devel..? btw I also found libgc (and I think gc-devel) with pacman.

**Comment by retsyo** [comment]
> are you sure you have libpcre installed, and not just pcre-devel..?

I can `make` the code stated in https://stackoverflow.com/questions/6509834/how-can-i-link-my-c-code-with-the-pcre-library-linker-errors-currently-being-t
And no matter whether I put `-lpcre` as the last parameter or not, I can't `make` minilight from shedskin

**Comment by retsyo** [comment]
I found the factor which yields `undefined reference to `__imp_pcre_malloc'`: since I prefer to static link, I add `-static` in `CPPFLAGS` environmental variant. If I remove `-static` from `CPPFLAGS`, then the code can be compiled, however, `libgc-1.dll`, `libgccpp-1.dll`, `libwinpthread-1.dll` and `libpcre-1.dll` should be suppiled with `exe`

**Comment by carlkl** [comment]
Adding `-DPCRE_STATIC ` will do the job IMHO.

**Comment by carlkl** [comment]
Sorry for opening again:

documentation: 

open an mingw32, mingw64 or ucrt64 shell ....  

(the msys shell is meant only for the Msys2 development)

**Comment by srepmub** [comment]
no problem, thanks for clarifying! I updated the instructions. please let me know if I made any other mistakes.
